flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/type/ExtendedSqlCollectionTypeNameSpec.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.sql.parser.type;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.SqlCollectionTypeNameSpec;
+import org.apache.calcite.sql.SqlTypeNameSpec;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.util.Litmus;
+import org.apache.calcite.util.Util;
+
+/**
+ * A extended sql type name specification of collection type,
+ * different with {@link SqlCollectionTypeNameSpec},
+ * we support NULL or NOT NULL suffix for the element type name(this syntax
+ * does not belong to standard SQL).
+ */
+public class ExtendedSqlCollectionTypeNameSpec extends SqlCollectionTypeNameSpec {
```

Comment:
Because the constructor of `SqlTypeNameSpec` constructor is package scope, previously i think Calcite's `SqlTypeNameSpec` is enough for Flink extension, but it's not, i have fixed this in CALCITE-3360, we can move them into `org.apache.flink.sql.parser.type` for the next Calcite version.

Reply:
ok, some comments will help reader to understand this.

Reply:
Sure, i have added a package-info file and address this info.